### ISSUE TITLE
Proposal: Tiller "Release Modules" to create pluggable backends

### ISSUE THREAD

**Issue by technosophos** [kubernetes/helm#1924]
# Proposal: Tiller Release Modules

A _release module_ is a pluggable Tiller component that can take a set of Kubernetes resource definitions and release them into a cluster. Release modules can replace Tiller's built in release mechanism, providing additional functionality or features.

- A release module is implemented as a gRPC server.
- A release module registers itself with Tiller
- `Chart.yaml` and `helm install` can instruct Tiller to use a particular release module
- When Tiller generates Kubernetes manifests, it will send them to the release module
- Tiller will then wait for the release module, and return the results to the user

## High-Level Summary

Normally, when Tiller receives an installation request, it does the following:

1. Read the chart
2. Calculate the values
3. Render the templates
4. Send the resource definitions to Kubernetes (default release module)
5. Return a result to the user

This proposal augments _step 4_ above by providing a mechanism for plugging in an alternative release module. A release module receives resource definitions and returns a status object indicating whether the release was successfully received (and, optionally, fully deployed) by Kubernetes.

- A Tiller server may have one or more release modules
- A chart (in Chart.yaml) may declare a specific release module that it needs to use
- The `helm install` command may specify which release module it needs to use
- If no release module is specified, Tiller may use whichever module is configured as its default
- If Tiller is incapable of installing a release with the requested release module, it _may_ fall back to the default

## Changes to Chart.yaml

The format for `Chart.yaml` will change in a backward-compatible way. The Chart.yaml will add the following fields:

```yaml
releaseModule:
  name: NAME
  options:
    KEY1: VALUE1
    KEY2: VALUE2
```

Here, `NAME` is the name of the release module, as registered on the server. `options` receives a map of `KEY`/`VALUE` pairs, where both key and value are strings.

The protobuf changes to `hapi.chart.Metadata` is as follows:

```proto
message Metadata {
	enum Engine {
		UNKNOWN = 0;
		GOTPL = 1;
	}
	// The name of the chart
	string name = 1;

   // ...
   // New fields:
   
	ReleaseModule releaseModule = 11;
}

message ReleaseModule {
    string name = 0;
    map<string, string> options = 1;
}
```

Note that no more than one `releaseModule` may be set per chart.

## Changes to the Helm Release Object

Helm release information is stored in a Helm Release (inside of a ConfigMap). This change will add a new field to the `hapi.release.Release` object:

```proto
message Release {
	// Name is the name of the release
	string name = 1;

   // ...
	// Namespace is the kubernetes namespace of the release.
	string namespace = 8;
	
	// New Fields
	hapi.chart.ReleaseModule releaseModule = 9;
}
```

While `releaseModule` will, in most cases, mirror the field sent from the client, if Tiller deploys using a different release module, it _must_ store that information here.

If no `releaseModule` is specified on a `Release` object, Tiller _must_ assume the default. This addresses migration and backward compatibility concerns.

## Changes to Helm/Tiller gRPC

The previous section outlines the changes to the protobuf that describes a Chart.yaml. This section focuses on gRPC calls.

### Install

Helm's `helm install` command will be augmented to provide a `--release-module=STRING` option that will allow users to supply or override the `releaseModule` setting from a Chart.yaml file.

The protobuf changes to `hapi.services.InstallReleaseRequest` are as follows:

```proto
message InstallReleaseRequest {
	// Chart is the protobuf representation of a chart.
	hapi.chart.Chart chart = 1;

    //...
    // New fields
    hapi.chart.ReleaseModule releaseModule = 10;
}
```

There is no change to the `InstallReleaseResponse`.

### Get

The Get operations may display the new data added to the `Release` object.

### Upgrade, Delete, Rollback

Upgrade, Delete, and Rollback must make an effort to use the same `releaseModule` that was used for installation.

As a special case, Delete _may_ fall back to the default release module.

Should the release module be unavailable, upgrade and rollback _must_ return an error, as the integrity of the release can no longer be guaranteed.

### Other Operations

No other operations are to be changed.

## Communication Between Tiller and the Release Module

This section describes how Tiller will operate release modules.

Tiller will connect to the release module over a gRPC connection in which Tiller will act as the client. Tiller will connect based on the endpoint specified by the module (see next section).

An implementation of a Release Module _must_ implement the following gRPC services:

- Install
- Upgrade
- Delete
- Rollback (which may function identically to Upgrade)

The complete protobuf spec for this is as follows:

```proto
syntax = "proto3";

package hapi.modules;

import "hapi/release/release.proto";

option go_package = "release";

service ReleaseModuleService {
	// InstallRelease requests installation of a chart as a new release.
	rpc InstallRelease(InstallReleaseRequest) returns (InstallReleaseResponse) {
	}

	// DeleteRelease requests deletion of a named release.
	rpc DeleteRelease(DeleteReleaseRequest) returns (DeleteReleaseResponse) {
	}

    // RollbackRelease rolls back a release to a previous version.
    rpc RollbackRelease(RollbackReleaseRequest) returns (RollbackReleaseResponse) {
    }

	// UpgradeRelease updates release content.
	rpc UpgradeRelease(UpgradeReleaseRequest) returns (UpgradeReleaseResponse) {
	}
}

message Result {
	enum Status {
		// No status set
		UNKNOWN = 0;
		// Operation was successful
		SUCCESS = 1;
		// Operation had no results (e.g. upgrade identical, rollback to same, delete non-existent)
		UNCHANGED = 2;
		// Operation failed
		ERROR = 3;
	}
	string info = 1;
	repeated string log = 2;
}

message InstallReleaseRequest {
	hapi.release.Release release = 1;
}
message InstallReleaseResponse {
	hapi.release.Release release = 1;
	Result result = 2;
}

message DeleteReleaseRequest {
	hapi.release.Release release = 1;
}
message DeleteReleaseResponse {
	hapi.release.Release release = 1;
	Result result = 2;
}

message UpgradeReleaseRequest{
	hapi.release.Release release = 1;
}
message UpgradeReleaseResponse{
	hapi.release.Release release = 1;
	Result result = 2;
}

message RollbackReleaseRequest{
	hapi.release.Release release = 1;
}
message RollbackReleaseResponse{
	hapi.release.Release release = 1;
	Result result = 2;
}
```

(Note that the names of these commands correspond with the Helm commands they pertain to.)

When Tiller receives an installation request, it will proceed as normal _until_ the templates have been successfully rendered. At that point, Tiller will have access to a rendered set of manifests, along with the original chart and installation objects.

At this point, Tiller will inspect Chart.yaml and the InstallRelease objects to find out what release module to use. In all cases, the InstallRelease information (if provided) is definitive. If no `InstallReleaseRequest.releaseModule` is specified, then the Chart's `releaseModule` section is used. If that is not specified, Tiller's default is used.

Tiller will verify that a release module with the given name exists. It will use that module or fall back to the default (or return an error).

Release modules must return a Release object object so that Helm can track the state of the release. Optionally, the Release object may be unchanged from the one sent to the release module.

### Hooks

Tiller will not execute hooks for a release module. A release module must handle hooks as appropriate.

## The Default Release Module

Tiller will include a pre-built in-process release module that performs the exact functionality that Tiller 2.2.0 supports. This module will be used when requested specifically, when a requested module is not present, or when no module is specified.

The default module will not implement the gRPC client/server pattern, but will run in-process.

## Registering a Release Server

Finally, a release server must be registered to Tiller. This is done by executing a gRPC request against a new endpoint on Tiller.

The protobuf message that must be sent to Tiler is described here, along with the description of how to delete a module from Tiller:

```proto
// ModuleRegistrationService provides a service for registering new modules to Tiller.
// This service is implemented by Tiller. Modules are clients of this service.
service ModuleRegistrationService {

	rpc AddModule(AddModuleRequest) returns (AddModuleResponse)

	rpc DeleteModule(DeleteModuleRequest) returns (DeleteModuleResponse)
}

message ModuleSpec {
	enum ModuleType {
		// Unset; should cause an error
		UNKNOWN = 0;
		// ReleaseModule
		RELEASE_MODULE = 1;
	}
	// Name is the name of the module. This must be globally unique.
	string name = 1;
	// Type is the module type.
	ModuleType type = 2;
	// Endpoint is the connection string for attaching a gRPC client to this module.
	string endpoint = 3;
}

message AddModuleRequest {
	ModuleSpec spec = 1
}
message AddModuleResponse {
	string msg = 1;
}

message DeleteModuleRequest {
	ModuleSpec spec = 1
}
message DeleteModuleResponse {
	string msg = 1;
}
```

For a release module, the `ModuleType` must be set to `RELEASE_MODULE`. In the future, Helm may support other module types.

NOTE: Tiller will store module registration information in config maps, and will read the configmaps at startup time.

## Backward Compatibility

Helm 2.x releases are guaranteed to be compatible in chart version formats, on-the-wire protocols (gRPC), template language, and file formats. Helm 2.x releases _should_ keep the public API of all accessible types, fields, and functions within `pkg/`, though there are occasions where this is not possible.

The following addresses various backward compatibility issues.

- Older versions of Tiller may install a chart with the default release module (the current release engine)
- The server must not require the presence of the new fields on the client requests, as an old Helm client must still be able to send requests
- The server must not require the presence of any new fields on the Chart.yaml

## Appendix: General Module Design

The design of the release module is intended to open a route for implementing other types of back-end modules. To that end, the module registration service provides a method of specifying modules of different types (though only `RELEASE_MODULE` is currently defined).

Additional modules would be added by:

- Adding new `ModuleType`s to the enumeration in `hapi.modules.ModuleSpec`
- Defining one or more new module services in `hapi.modules`
- Exposing the modules to the client when appropriate

**Comment by pigmej** [comment]
cc @nebril

**Comment by TamalSaha** [comment]
Hi,
I have major reservation about how proposal allows modifying the gut of Tiller. I see the value of extensibility but IMHO the downsides are many.

Fragmentation of Charts: We would like Helm project to define and tightly control the UX for application deployment. But that breaks down with this proposal. Now, someone can write a Chart that works with a particular module. This will be bad for UX when a user finds out that  a particular Chart from some public repo does not work with the default Tiller. To me this is akin to doing a docker pull and finding out that the image format does not work with default Docker. This is unacceptable IMHO.

This introduces all the complexity of UAS in the Kubernetes aggregate api server discussion inside Tiller. https://groups.google.com/forum/#!msg/kubernetes-sig-apps/0gbmMNvZWUo/fgYSHNoWCQAJ 

This proposal will introduce significant complexity for cluster operators. We would like to run a hosted version of Tiller. But this proposal means that that will be not a meaningful pursuit. Because

- We may have to run multiple controller.
- We may not have access to each module implementation for licensing and other purposes. 

The auth information has to be propagated to make things work here. 

If the goal is allow experimentation, then my proposal will be to just tell people to write their own Tiller. The new service defined in this proposal is almost same as the main Tiller service. New implementors might as well, implement the full interface using existing Tiller rendering code as a library. And Helm cli already supports setting --host.

I agree Tiller/Helm needs to support a better DAG for complex deployments. I think the right way to do that will be to add to the core. This will mean barriers for entry for new executors. I think that is needed to avoid fragmentation and ultimate success for Helm.

**Comment by nebril** [comment]
Hi,

@technosophos Do you think there is a possibility for two different charts to use two different backends in a single release? I think such situation shouldn't be allowed, how do we mitigate it?

@TamalSaha what if modules themselves were charts and could be deployed in each cluster "automagically"?

I can see that happening to AppController.

The only problem I see is that these charts could not be dependencies for a release they are supposed to orchestrate because they would need to be available to deploy themselves, or the backend would switch mid-release to the new backend, which I imagine would be bad and hard to implement, not mentioning debugging in production.

Overall, I like the proposal, the one think I would like to add is that Release Servers should be deployable as charts to help with UX.

**Comment by pigmej** [comment]
Maybe instead we should somehow define a subset of features that Release Module has to support. It would behave then in similar way as CRI, where you have subset of features that you *have to* support, and all "unique" features for each runtime implemenation is still working (outside of CRI though). I know that this proposal describes such interface but I think allowing selection of *given* module per chart is bad assumption. Chart should be always deployed with configured / enabled Release Module.

Then problem with multiple modules would be resolved, fragmentation would be also solved (all Release Modules would have to support the same operations).

Those are mine 3c.

**Comment by technosophos** [comment]
On option would be to simplify the system this way:

- Tiller only allows _one_ release backend
- That backend must handle all deployments
- The client has no way of specifying a backend

In that way, it'd basically make it possible to replace the default deployment logic, but only with a single backend that would be required to handle any chart.

Does that get us any closer to the desired feature set?

**Comment by pigmej** [comment]
I think it allows us to do some useful stuff. It will be more or less equal to CRI. @nebril what do you think?

In case of AC it should allow us to integrate, and still add functionality to Helm.

**Comment by nebril** [comment]
@technosophos how would activating release backend work in the simplified case? Would it be set as an argument during `helm init`?

**Comment by TamalSaha** [comment]
I just saw this: 
http://blog.slashdeploy.com/2017/02/16/introducing-await/

**Comment by pigmej** [comment]
@TamalSaha how it's related? Await is "just" a cli tool which waits for condition ;) It's a tool that could be used in init-containers to kinda orchestrate some stuff, but this approach is hard to debug and see dependencies between components.

**Comment by TamalSaha** [comment]
@pigmej I was not thinking about their solution. I was point to the fact that they were having issues AppController trying to address in their Helm chart.

**Comment by pigmej** [comment]
ah got it :)

**Comment by TamalSaha** [comment]
@pigmej , @nebril Have you got a chance to look at the latest comment on Auth work? https://github.com/kubernetes/helm/issues/1918#issuecomment-280970015

Thinks got complicated . This affect how auth works in a backend module context.

**Comment by technosophos** [comment]
Summary of the Tiller Backend Working Group

Agreed that:

- Any satisfactory backend must be able to deploy any standard chart
- We wanted to support **just one** release backend at a time
- CLI/Chart should not be able to "specify a backend". Tiller will always use the one configured backend

But we still have open issues on:

- Should the backend be compiled into Tiller (like Storage Backends), or should it be a separate microservice (as originally proposed above)?
  - If we want it compiled in, we need to write a Go interface like the one for storage backends (https://github.com/kubernetes/helm/blob/master/pkg/storage/driver/driver.go)
  - If we want to stick with a gRPC interface, what do we need to modify in the original proposal above?
- To what extend do we need to worry about AuthN/AuthZ for...
  - Communication between Tiller and a (possible) backend
  - Verifying that a user is allowed to create the TPRs for AppController
  - Validating inside of AppController (or any backend plugin) that it can trust the user who sent the request
- Is the backend specified at startup time (for example, by environment variables in Tiller's Deployment manifeste)? Or is it configured at runtime via an API (as in the original proposal above)?
- Can this be done in the 2.3.0 timeframe?

**Comment by pigmej** [comment]
> To what extend do we need to worry about AuthN/AuthZ for...

I think we may use Kubernetes to solve that problem by spawning Tiler and module in single pod and then using localhost to communicate. Then all AuthN/AuthZ is still in Tiler, and it's still safe to use it. 
(that was just fast thought)

> Can this be done in the 2.3.0 timeframe?

I strongly believe that it's possible :)

**Comment by TamalSaha** [comment]
> Validating inside of AppController (or any backend plugin) that it can trust the user who sent the request

If backend runs as a compiled binary, then Tiller's auth will take care of it. If backend modules run as separate gRPC services, then one option will be to use the Tiller's[ auth proxy option](https://github.com/kubernetes/helm/issues/1918#issuecomment-280970015). This is the recommended approach for compatibility for UAS-es. Basically there is a separate client cert auth happens between Tiller and Backend module. Then Tiller can forward the `userInfo` to Backend module to impersonate.

**Comment by TamalSaha** [comment]
@pigmej / @nebril  Some questions about AppController:

- I think more clarifications are needed for `Any satisfactory backend must be able to deploy any standard chart`. If all backends are "same", then why will users use a different backend? I noticed that AppController introduces new labels for Deployments/ RCs like ike  [`success_factor`](https://github.com/Mirantis/k8s-AppController/blob/adcf6e48222d17c01f150a1626fdc46267338525/README.md#dependencies). Is every implementation of backend module need to support options like [`success_factor`] ? The other way to look at this question is what is open for experimentation for Backend modules?

- I saw that AppController has a hard coded list of resources in its [Definition](https://github.com/Mirantis/k8s-AppController/blob/cfc62d2de68d9e7f52c9c84f976dc93502bf59ab/pkg/client/resdef.go#L42) TPR. How does it deal with objects it does not know?

Thanks.

**Comment by pigmej** [comment]
@TamalSaha 

ad1. AC offers much more than `success_factor`. And it will have more and more stuff. It's kinda obvious that when you create unified interface on top of something, you're pretty much bound to minimal shared subset of functionalities.

ad2. We will need to support all of objects ;-) Simple as that. AC shouldn't limit Helm possibilities, AC and any other module *have to* support all objects supported by Helm.

**Comment by TamalSaha** [comment]
> ad1. AC offers much more than success_factor. And it will have more and more stuff. It's kinda obvious that when you create unified interface on top of something, you're pretty much bound to minimal shared subset of functionalities

I think you get the center of my main concern: `fragmentation`. If some charts need functionality to properly install apps that are not available in the "minimal shared subset of functionalities", will those Charts be accepted in the https://github.com/kubernetes/charts repo? 

Also, how does Tiller's default implementation evolve over time? Given dependency detection is implicit, one can send a patch to support AC like dependency management in the default implementation. Will a patch like that be accepted?

**Comment by pigmej** [comment]
Well, no ‎the integration will be limited to 'common subset'. It's the same for any library which supports different backends for doing xyz, same for cri, same for any db implementation etc.

All unique features will be not supported in this integration. User will need to enable them later in post deployment or sth. 

Adding patch that does dependency 'in helm directly' is I think out of scope for helm 'core'. Maybe there should be default / preffered implementation that comes from helm subrepo? I don't see a benefit of hardcoding dependency management into helm right now especially that this logic is currently *not* the part of helm.

**Comment by TamalSaha** [comment]
> Adding patch that does dependency 'in helm directly' is I think out of scope for helm 'core'. Maybe there should be default / preffered implementation that comes from helm subrepo? I don't see a benefit of hardcoding dependency management into helm right now especially that this logic is currently *not* the part of helm.

I disagree. If this is a common enough problem for Helm's users, then Helm core is ought to solve in its default implementation.

**Comment by TamalSaha** [comment]
@pigmej , I think you missed my question. Will any chart that uses "unique" features of a backend be accepted into the official Chart repo?

**Comment by pigmej** [comment]
@TamalSaha no unique features shouldn't exists from Helm point of view.

**Comment by pigmej** [comment]
@TamalSaha also release module in theory may do more than just during "deployment". I'm speaking about some leader reelection stuff, or mysql-multi-slave and slave propagation etc. These actions are unknown for helm and they probably should remain unknown/out of scope.

**Comment by TamalSaha** [comment]
> @TamalSaha also release module in theory may do more than just during "deployment". I'm speaking about some leader reelection stuff, or mysql-multi-slave and slave propagation etc. These actions are unknown for helm and they probably should remain unknown/out of scope.

Agree. I will again say that the pure dependency feature belongs in the core.

**Comment by nebril** [comment]
@TamalSaha what do you mean by "pure dependency"? Dependencies within the charts? I think this can be discussed here, as it is connected. In the end though we will need a whole new proposal which will have e.g. chart format changes (for specifying dependencies).

**Comment by TamalSaha** [comment]
I meant the type of dependency AppController is handling. That should be handled in core as all the information is already in Chart.

**Comment by nebril** [comment]
AppController could handle dependencies in chart and between charts. So I am not sure what should be handled in core in your opinion.

**Comment by TamalSaha** [comment]
Ideally both "dependencies in chart and between charts".

**Comment by TamalSaha** [comment]
The issue is Helm should decide what are necessary minimal features that should come with Helm (as the default feature set). Those features can be implemented as backend modules (same binary, separate grpc service). That is a technical question and my vote is for simplicity, hence same binary.

But the bigger question is these minimal necessary features sets needs to be part of the Helm community project (under CNCF). It could be same git or separate git repo. We, AppsCode, we like to know that our investment in Helm is "safe" and does not dependent on business decisions on entities external to Helm project. 2) As things evolve the "set of minimal necessary features" can evolve on customer demand.

**Comment by nebril** [comment]
@TamalSaha AppController is going to join K8s incubator soon, with aim of becoming official K8s project. I hope this will ensure you that it's "safe" to use.

**Comment by TamalSaha** [comment]
@nebril That is good to hear!

Assuming Tiller takes a nginx/apache module approach, how does new modules are added to Helm without each module having to become a Kubernetes incubator project?

**Comment by technosophos** [comment]
Okay, below is an attempt to substantially revise the proposal. This one still uses gRPC, which currently still seems to be easier to deal with (from a code-management POV) then compiling the module directly in.

# Proposal: Tiller Release Modules

This document describes a plugin model for Tiller. Specifically, it defines a modular interface for changing the _release (deployment) backend_. This is intended to be a first step toward making Tiller modular in the same way that Apache or nginx are modular.

A _release module_ is a pluggable Tiller component that can take a set of Kubernetes resource definitions and release them into a cluster. Release modules can replace Tiller's built in release mechanism, providing additional functionality or features.

- A release module is implemented as a gRPC server.
- A release module registers itself with Tiller
- During a `helm` install, upgrade, delete, or rollback request...
	- When Tiller generates Kubernetes manifests, it will send them to the release module
	- Tiller will then wait for the release module, and return the results to the user

**This feature will be deemed _experimental_, and will be subject to future API-breaking changes until the experimental flag is removed.**

## High-Level Summary

Normally, when Tiller receives an installation request, it does the following:

1. Read the chart
2. Calculate the values
3. Render the templates
4. Send the resource definitions to Kubernetes (default release module)
5. Return a result to the user

This proposal augments _step 4_ above by providing a mechanism for plugging in an alternative release module. A release module receives resource definitions and returns a status object indicating whether the release was successfully received (and, optionally, fully deployed) by Kubernetes.

## Communication Between Tiller and the Release Module

This section describes how Tiller will operate release modules.

Tiller will connect to the release module over a gRPC connection in which Tiller will act as the client. Tiller will connect based on the endpoint specified by the module (see next section).

An implementation of a Release Module _must_ implement the following gRPC services:

- Install
- Upgrade
- Delete
- Rollback (which may function identically to Upgrade)

The complete protobuf spec for this is as follows:

```proto
syntax = "proto3";

package hapi.modules;

import "hapi/release/release.proto";

option go_package = "release";

service ReleaseModuleService {
	rpc Version(VersionReleaseRequest) return (VersionReleaseResponse)  {
	}

	// InstallRelease requests installation of a chart as a new release.
	rpc InstallRelease(InstallReleaseRequest) returns (InstallReleaseResponse) {
	}

	// DeleteRelease requests deletion of a named release.
	rpc DeleteRelease(DeleteReleaseRequest) returns (DeleteReleaseResponse) {
	}

    // RollbackRelease rolls back a release to a previous version.
    rpc RollbackRelease(RollbackReleaseRequest) returns (RollbackReleaseResponse) {
    }

	// UpgradeRelease updates release content.
	rpc UpgradeRelease(UpgradeReleaseRequest) returns (UpgradeReleaseResponse) {
	}
}

message Result {
	enum Status {
		// No status set
		UNKNOWN = 0;
		// Operation was successful
		SUCCESS = 1;
		// Operation had no results (e.g. upgrade identical, rollback to same, delete non-existent)
		UNCHANGED = 2;
		// Operation failed
		ERROR = 3;
	}
	string info = 1;
	repeated string log = 2;
}

message VersionReleaseRequest {
}

message VersionReleaseResponse {
   string name = 1;     // The canonical name of the release module
   string version = 2;  // The version of the release module
}

message InstallReleaseRequest {
	hapi.release.Release release = 1;
}
message InstallReleaseResponse {
	hapi.release.Release release = 1;
	Result result = 2;
}

message DeleteReleaseRequest {
	hapi.release.Release release = 1;
}
message DeleteReleaseResponse {
	hapi.release.Release release = 1;
	Result result = 2;
}

message UpgradeReleaseRequest{
	hapi.release.Release release = 1;
}
message UpgradeReleaseResponse{
	hapi.release.Release release = 1;
	Result result = 2;
}

message RollbackReleaseRequest{
	hapi.release.Release release = 1;
}
message RollbackReleaseResponse{
	hapi.release.Release release = 1;
	Result result = 2;
}
```

(Note that the names of these commands correspond with the Helm commands they pertain to.)

When Tiller receives an installation request, it will proceed as normal _until_ the templates have been successfully rendered. At that point, Tiller will have access to a rendered set of manifests, along with the original chart and installation objects.

At this point, if a release module is activated, Tiller will initialize a gRPC client and forward the request to the release module. (If no release module is activated, the default in-process backend will be used.)

Release modules _must_ return a Release object object so that Helm can track the state of the release. Optionally, the Release object may be unchanged from the one sent to the release module.

_If a release module is configured, but is unresponsive, Tiller **must** return an error._

### Hooks

Tiller will not execute hooks for a release module. A release module must handle hooks as appropriate.

## The Default Release Module

Tiller will include a pre-built in-process release module that performs the exact functionality that Tiller 2.2.0 supports. This module will be used when requested specifically, when a requested module is not present, or when no module is specified.

The default module will not implement the gRPC client/server pattern, but will run in-process.

## Registering a Release Module via Tiller Environment Variables

For the initial release of the release module capability, a release module will be registered via environment variables passed in through the :

```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  creationTimestamp: null
  labels:
    app: helm
    name: tiller
  name: tiller-deploy
  namespace: kube-system
spec:
  replicas: 1
  strategy: {}
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: helm
        name: tiller
    spec:
      containers:
        image: gcr.io/kubernetes-helm/tiller:v2.2.0
        imagePullPolicy: IfNotPresent
        - env:
          - name: TILLER_NAMESPACE
            value: kube-system
          - name: RELEASE_MODULE_NAME
            value: MyReleaseModule
          - name: RELEASE_MODULE_ENDPOINT
            value: 10.21.77.44:4434
```

The above adds two variables:

- `RELEASE_MODULE_NAME` should be the name of the backend module, and will be used for logging and other identification purposes
- `RELEASE_MODULE_ENDPOINT` should be the gRPC endpoint of the service

## Modifications to the `{{ .Capabilities }}` Template Object

The `.Capabilities` object that is accessible inside of templates will be extended to include this object:

```
.Capabilities.Module.Release.Name = $NAME
.Capabilities.Module.Release.Version = $SEMVER
```

Where `$NAME` will be the name of the module (default: `default`) and `$SEMVER` will be the SemVer 2 compliant version number of the module (default: the Tiller version (`2.3.0`).

This will allow template authors access to information about the release backend. (The structure of this field is designed to accommodate future modules.)

## Compatibility

An implementation of the release backend must be able to install any charts that the default backend can install.

## Non-Goals

For the first implementation, we will not be able to implement an authentication mechanism because this feature set is currently not implemented. But since this is marked experimental, we will reserve the right to significantly modify this portion later, as needs dictate.

## APPENDIX: (DEFERRED) Registering a Release Module over gRPC

_This section reflects a possible future direction, and is not assumed to be part of the initial build_

Finally, a release server must be registered to Tiller. This is done by executing a gRPC request against a new endpoint on Tiller.

The protobuf message that must be sent to Tiler is described here, along with the description of how to delete a module from Tiller:

```proto
// ModuleRegistrationService provides a service for registering new modules to Tiller.
// This service is implemented by Tiller. Modules are clients of this service.
service ModuleRegistrationService {

	rpc AddModule(AddModuleRequest) returns (AddModuleResponse)

	rpc DeleteModule(DeleteModuleRequest) returns (DeleteModuleResponse)
}

message ModuleSpec {
	enum ModuleType {
		// Unset; should cause an error
		UNKNOWN = 0;
		// ReleaseModule
		RELEASE_MODULE = 1;
	}
	// Name is the name of the module. This must be globally unique.
	string name = 1;
	// Type is the module type.
	ModuleType type = 2;
	// Endpoint is the connection string for attaching a gRPC client to this module.
	string endpoint = 3;
}

message AddModuleRequest {
	ModuleSpec spec = 1
}
message AddModuleResponse {
	string msg = 1;
}

message DeleteModuleRequest {
	ModuleSpec spec = 1
}
message DeleteModuleResponse {
	string msg = 1;
}
```

For a release module, the `ModuleType` must be set to `RELEASE_MODULE`. In the future, Helm may support other module types.

NOTE: Tiller will store module registration information in config maps, and will read the configmaps at startup time.

**Comment by bacongobbler** [comment]
From an outsider's perspective with no idea what this is trying to solve, what does this proposal accomplish for the end-user? Just trying to get some bearings on the issue at hand to see if this is relevant to my efforts or not.

**Comment by nebril** [comment]
@bacongobbler currently helm installs a release by grabbing all object manifests and creating them in k8s. This is fine for stateless apps, but not all applications are stateless. This proposal is about figuring out how to define pluggable release backends which will install a release that needs more sophisticated orchestration - like waiting for dependency charts to be ready.

**Comment by TamalSaha** [comment]
`.Capabilities` object will allow template authors access to information about the release backend. This means, template authors can write Charts that only work for a given module. On the other hand Tiller can have only one backend. The combination of this two will mean users will not be able to run 2 Charts that need 2 separate modules. And this also opens up path for fragmentation of Charts. _I think this information should not be exposed in the Capabilities._

Regarding authentication, Tiller and it's backend module can run side by side and use localhost. This will solve the problem of authentication for now.

**Comment by TamalSaha** [comment]
I just dawned on me that, may be an `.Experimental` node can be added that has  `.Capabilities` under it.
```
.Experimental.Capabilities.Module.Release.Name = $NAME
.Experimental..Capabilities.Module.Release.Version = $SEMVER
```
The "contract" is that no Chart will be accepted in kubernetes/charts that uses  `.Experimental.` features. Once there is enough evidence/use-case/understanding of an experimental feature it can be promoted from `experimental`. Then any chart using that can also be merged into official repos.

May be that can unblock everyone.

**Comment by nebril** [comment]
There is another method that Rudder will need to implement, it's release status check. AppController (or other orchestrator) need to check the status of a release and map it to tiller release status, so that helm user can see what is happening with his release. It will be done in #2079

**Comment by wstrange** [comment]
Playing armchair architect here :-)  

Related to @bacongobbler question (what does this do for the end user?) and @TamalSaha concerns,  this proposal seems to introduce a lot of complexity for very little gain.  A remote service is one more moving part that could go wrong.

From what I can see, there is significant pain around workflow and orchestration. Many folks are coming at this from different angles - but the underlying concern is the same.

I can see how a pluggable rudder could solve workflow - but then again, if this is so important (and I think it is),  would it not be better to make orchestration a core part of helm/tiller?  

I share @TamalSaha concern that pluggable backends could lead to fragmentation of helm.  It would be one more piece of environment specific configuration that I need to ensure is in place before I can deploy my application.

**Comment by nebril** [comment]
@wstrange: on one hand most k8s applications don't need orchestration.

on the other hand when you have an app that needs it and you want to use helm to deploy it you need custom tiller, which is PITA.

When this proposal is implemented, it gives you a possibility to swap out your orchestration backend, and only it. It doesn't affect current user experience at all.

I don't agree about the moving parts argument, the number of parts is mostly the same, we are just extracting them to a new microservice which becomes a swappable part.

**Comment by technosophos** [comment]
Marking this as closed, since the Rudder implementation has been merged
